# FG10_alpha_fold

## Introduction and Overview
This repository contains scripts for extracting key structural information from alpha-fold protein models. The extracted data includes atomic site properties and structural conformation features.

## Directory Structure
The directory structure is as follows:

```bash
DrivR-Base/
|-- FG10_alpha_fold/
|   |-- get_alpha_fold_scores_atom.sh
|   |-- get_alpha_fold_atom.py
|   |-- get_alpha_fold_scores_struct_conform.sh
|   |-- get_alpha_fold_struct_conf.py
|   |-- package_dependencies.py
```
## Packages and Dependencies
Package dependencies are managed through the **package_dependencies.py** script. 

## Prerequisite Data
To extract information from the CIF alpha fold file, you need details about the protein variant's position and impact. This information is derived from the '**_variant_effect_output_all.txt**' results, generated by the '**FG2_vep/query_vep.sh script**'. However, if you wish to use this as a stand-alone module then the input file should follow the format below:

| chrom |   pos  | ref_allele | alt_allele |  R  | driver_stat |    gene    | protein_position |
| ----- | ------ | ---------- | ---------- | --- | ----------- | ---------- | ---------------- |
| chr1  | 935778 |     C      |      T     |  1  |      1      |   SAMD11   |        28        |

## Script Usage
The two main scripts are **get_alpha_fold_atom.py** and **get_alpha_fold_struct_conf.py**. Both scripts require the amino acid query results (**'variants + "_variant_effect_output_all.bed"'**) from VEP. If using this script as a standalone module, format the input file as described above, replace the filename in the for **'chunk in pd.read_csv(...)'** part with the appropriate name, and assign **'chunk'** to **'df3'**.

**get_alpha_fold_atom.py:**

* Extracts five atomic features for a given protein position:
  *  [**X Atom-Site Coordinate**](https://mmcif.wwpdb.org/dictionaries/mmcif_pdbx_v40.dic/Items/_atom_site.Cartn_x.html)
  *  [**Y Atom-Site Coordinate**](https://mmcif.wwpdb.org/dictionaries/mmcif_pdbx_v40.dic/Items/_atom_site.Cartn_y.html)
  *  [**Z Atom-Site Coordinate**](https://mmcif.wwpdb.org/dictionaries/mmcif_pdbx_v40.dic/Items/_atom_site.Cartn_z.html)
  *  [**Atom Site Occupancy**](https://mmcif.wwpdb.org/dictionaries/mmcif_rcsb_nmr.dic/Items/_atom_site.occupancy.html)
  *  [**Isotropic Temperature Factor Parameter**](https://mmcif.wwpdb.org/dictionaries/mmcif_rcsb_nmr.dic/Items/_atom_site.B_iso_or_equiv.html)

**get_alpha_fold_struct_conf.py:**

* Extracts [**structural conformation features**](http://mmcif.rcsb.org/dictionaries/mmcif_pdbx_v31.dic/Categories/struct_conf.html) features for a given protein position.
* Features are based on imputations from PDB and include:
  *  [**Structural Conformation Type ID**](https://mmcif.wwpdb.org/dictionaries/mmcif_pdbx_v50.dic/Items/_struct_conf.conf_type_id.html)
  *  [**Structural Conformation ID**]([bx_v50.dic/Items/_struct_conf.id.html](https://mmcif.wwpdb.org/dictionaries/mmcif_pdbx_v50.dic/Items/_struct_conf.id.html)https://mmcif.wwpdb.org/dictionaries/mmcif_pdbx_v50.dic/Items/_struct_conf.id.html)
 
## Conclusion
These scripts facilitate the extraction of crucial atomic and structural information from alpha fold protein models, aiding in the analysis of protein variants' effects.
